File: backend/internal/dto/delivery.go
package dto

import (
	"time"

	"github.com/shopspring/decimal"
)

type DeliveryOrderCreateRequest struct {
	OrderID         uint   `json:"order_id" binding:"required"`
	DeliveryType    string `json:"delivery_type" binding:"required,oneof=self meituan eleme"`
	SenderName      string `json:"sender_name"`
	SenderPhone     string `json:"sender_phone"`
	SenderAddress   string `json:"sender_address"`
	SenderLng       float64 `json:"sender_lng"`
	SenderLat       float64 `json:"sender_lat"`
	ReceiverName    string `json:"receiver_name" binding:"required"`
	ReceiverPhone   string `json:"receiver_phone" binding:"required"`
	ReceiverAddress string `json:"receiver_address" binding:"required"`
	ReceiverLng     float64 `json:"receiver_lng"`
	ReceiverLat     float64 `json:"receiver_lat"`
}

type DeliveryOrderResponse struct {
	ID              uint            `json:"id"`
	OrderID         uint            `json:"order_id"`
	OrderNo         string          `json:"order_no"`
	StoreID         uint            `json:"store_id"`
	DeliveryType    string          `json:"delivery_type"`
	DeliveryStatus  int             `json:"delivery_status"`
	RiderID         uint            `json:"rider_id"`
	RiderName       string          `json:"rider_name"`
	RiderPhone      string          `json:"rider_phone"`
	RiderLng        float64         `json:"rider_lng"`
	RiderLat        float64         `json:"rider_lat"`
	DeliveryFee     decimal.Decimal `json:"delivery_fee"`
	Distance        float64         `json:"distance"`
	Duration        int             `json:"duration"`
	SenderName      string          `json:"sender_name"`
	SenderPhone     string          `json:"sender_phone"`
	SenderAddress   string          `json:"sender_address"`
	ReceiverName    string          `json:"receiver_name"`
	ReceiverPhone   string          `json:"receiver_phone"`
	ReceiverAddress string          `json:"receiver_address"`
	PlatformOrderID string          `json:"platform_order_id"`
	PlatformType    string          `json:"platform_type"`
	EstimatedTime   *time.Time      `json:"estimated_time"`
	PickedUpAt      *time.Time      `json:"picked_up_at"`
	DeliveredAt     *time.Time      `json:"delivered_at"`
	CreatedAt       time.Time       `json:"created_at"`
}

type PickupCodeResponse struct {
	OrderID   uint   `json:"order_id"`
	Code      string `json:"code"`
	Status    int    `json:"status"`
	ExpiredAt string `json:"expired_at,omitempty"`
}

type RiderLocationUpdate struct {
	RiderID uint    `json:"rider_id" binding:"required"`
	Lng     float64 `json:"lng" binding:"required"`
	Lat     float64 `json:"lat" binding:"required"`
	Speed   float64 `json:"speed"`
	Heading float64 `json:"heading"`
}

type RiderLocationResponse struct {
	RiderID   uint    `json:"rider_id"`
	RiderName string  `json:"rider_name"`
	Lng       float64 `json:"lng"`
	Lat       float64 `json:"lat"`
	Speed     float64 `json:"speed"`
	Heading   float64 `json:"heading"`
	UpdatedAt string  `json:"updated_at"`
}

type RoutePlanRequest struct {
	OriginLng float64 `json:"origin_lng" binding:"required"`
	OriginLat float64 `json:"origin_lat" binding:"required"`
	DestLng   float64 `json:"dest_lng" binding:"required"`
	DestLat   float64 `json:"dest_lat" binding:"required"`
}

type RoutePlanResponse struct {
	Distance float64 `json:"distance"`
	Duration int     `json:"duration"`
	Route    string  `json:"route"`
	Fee      decimal.Decimal `json:"fee"`
}

type GeocodeRequest struct {
	Address string `json:"address" binding:"required"`
	City    string `json:"city"`
}

type GeocodeResponse struct {
	Lng       float64 `json:"lng"`
	Lat       float64 `json:"lat"`
	Formatted string  `json:"formatted"`
}

type CreateRiderRequest struct {
	Name  string `json:"name" binding:"required"`
	Phone string `json:"phone" binding:"required"`
}

type DeliveryOrderQuery struct {
	PageQuery
	StoreID        uint   `form:"store_id"`
	DeliveryType   string `form:"delivery_type"`
	DeliveryStatus int    `form:"delivery_status"`
	OrderID        uint   `form:"order_id"`
}

type DeliveryStatusUpdateRequest struct {
	DeliveryStatus int `json:"delivery_status" binding:"required"`
}

type AssignRiderRequest struct {
	RiderID uint `json:"rider_id" binding:"required"`
}

type VerifyPickupCodeRequest struct {
	Code    string `json:"code" binding:"required"`
	StoreID uint   `json:"store_id" binding:"required"`
}

type DeliveryTrackingResponse struct {
	OrderNo         string               `json:"order_no"`
	DeliveryType    string               `json:"delivery_type"`
	DeliveryStatus  int                  `json:"delivery_status"`
	RiderID         uint                 `json:"rider_id"`
	RiderName       string               `json:"rider_name"`
	RiderPhone      string               `json:"rider_phone"`
	RiderLng        float64              `json:"rider_lng"`
	RiderLat        float64              `json:"rider_lat"`
	Distance        float64              `json:"distance"`
	Duration        int                  `json:"duration"`
	ReceiverAddress string               `json:"receiver_address"`
	SenderAddress   string               `json:"sender_address"`
	EstimatedTime   *time.Time           `json:"estimated_time"`
	Trackings       []TrackingPoint      `json:"trackings"`
}

type TrackingPoint struct {
	Lng       float64 `json:"lng"`
	Lat       float64 `json:"lat"`
	Speed     float64 `json:"speed"`
	Timestamp int64   `json:"timestamp"`
}

File: miniprogram/src/services/delivery.ts
import request from './request'

export type DeliveryType = 'self' | 'meituan' | 'eleme'
export type OrderType = 'dine_in' | 'takeout' | 'delivery' | 'pickup'

export interface DeliveryOrder {
  id: number
  order_id: number
  order_no: string
  store_id: number
  delivery_type: DeliveryType
  delivery_status: number
  rider_id: number
  rider_name: string
  rider_phone: string
  rider_lng: number
  rider_lat: number
  delivery_fee: number
  distance: number
  duration: number
  sender_name: string
  sender_phone: string
  sender_address: string
  receiver_name: string
  receiver_phone: string
  receiver_address: string
  platform_order_id: string
  platform_type: string
  estimated_time: string | null
  picked_up_at: string | null
  delivered_at: string | null
  created_at: string
}

export interface DeliveryTracking {
  order_no: string
  delivery_type: DeliveryType
  delivery_status: number
  rider_id: number
  rider_name: string
  rider_phone: string
  rider_lng: number
  rider_lat: number
  distance: number
  duration: number
  receiver_address: string
  sender_address: string
  estimated_time: string | null
  trackings: TrackingPoint[]
}

export interface TrackingPoint {
  lng: number
  lat: number
  speed: number
  timestamp: number
}

export interface RiderLocation {
  rider_id: number
  rider_name: string
  lng: number
  lat: number
  speed: number
  heading: number
  updated_at: string
}

export interface PickupCodeInfo {
  order_id: number
  code: string
  status: number
  expired_at?: string
}

export interface RoutePlanResult {
  distance: number
  duration: number
  route: string
  fee: number
}

export interface GeocodeResult {
  lng: number
  lat: number
  formatted: string
}

export const deliveryStatusMap: Record<number, { text: string; color: string }> = {
  0: { text: '待接单', color: '#e6a23c' },
  1: { text: '骑手已接单，取餐中', color: '#409eff' },
  2: { text: '配送中', color: '#409eff' },
  3: { text: '已送达', color: '#67c23a' },
  4: { text: '已取消', color: '#909399' },
}

export const orderTypeMap: Record<OrderType, { label: string; icon: string }> = {
  dine_in: { label: '堂食', icon: '🍽️' },
  takeout: { label: '外带', icon: '🥡' },
  delivery: { label: '外卖配送', icon: '🛵' },
  pickup: { label: '到店自提', icon: '🏪' },
}

export const createDeliveryOrder = (data: {
  order_id: number
  delivery_type: DeliveryType
  receiver_name: string
  receiver_phone: string
  receiver_address: string
  receiver_lng?: number
  receiver_lat?: number
  sender_name?: string
  sender_phone?: string
  sender_address?: string
  sender_lng?: number
  sender_lat?: number
}) => {
  return request<DeliveryOrder>({
    url: '/delivery',
    method: 'POST',
    data
  })
}

export const getDeliveryByOrder = (orderId: number) => {
  return request<DeliveryOrder>({
    url: `/delivery/order/${orderId}`,
    method: 'GET',
    needLogin: false
  })
}

export const getDeliveryTracking = (orderId: number) => {
  return request<DeliveryTracking>({
    url: `/delivery/tracking/${orderId}`,
    method: 'GET',
    needLogin: false
  })
}

export const getRiderLocation = (riderId: number) => {
  return request<RiderLocation>({
    url: `/riders/${riderId}/location`,
    method: 'GET'
  })
}

export const generatePickupCode = (orderId: number, storeId: number) => {
  return request<PickupCodeInfo>({
    url: '/pickup/code',
    method: 'POST',
    data: { order_id: orderId, store_id: storeId }
  })
}

export const verifyPickupCode = (code: string, storeId: number) => {
  return request<PickupCodeInfo>({
    url: '/pickup/verify',
    method: 'POST',
    data: { code, store_id: storeId }
  })
}

export const getPickupCodeByOrder = (orderId: number) => {
  return request<PickupCodeInfo>({
    url: `/pickup/order/${orderId}`,
    method: 'GET',
    needLogin: false
  })
}

export const planRoute = (originLng: number, originLat: number, destLng: number, destLat: number) => {
  return request<RoutePlanResult>({
    url: '/amap/route',
    method: 'POST',
    data: { origin_lng: originLng, origin_lat: originLat, dest_lng: destLng, dest_lat: destLat }
  })
}

export const geocode = (address: string, city?: string) => {
  return request<GeocodeResult>({
    url: '/amap/geocode',
    method: 'POST',
    data: { address, city }
  })
}

export const simulateRiderLocation = (deliveryId: number) => {
  return request<{ message: string }>({
    url: `/delivery/${deliveryId}/simulate-location`,
    method: 'POST'
  })
}
